lib/patchelf/alt_saver.rb

```diff
@@ -622,42 +635,169 @@ def rewrite_sections_library
+        size = 0
+        sections_at_aligned_offset(curr_off) do |sec|
+          next if sec.type != ELFTools::Constants::SHT_NOTE
+
+          size = sec.header.sh_size.to_i
+          curr_off = sec.header.sh_offset.to_i
+          break
+        end
+
+        raise PatchError, 'cannot normalize PT_NOTE segment: non-contiguous SHT_NOTE sections' if size.zero?
+
+        if curr_off + size > end_off
+          raise PatchError, 'cannot normalize PT_NOTE segment: partially mapped SHT_NOTE section'
+        end
+
+        new_phdr = ELFTools::Structs::ELF_Phdr[elf_class].new(endian: endian, **phdr.snapshot)
+        new_phdr.p_offset = curr_off
+        new_phdr.p_vaddr = phdr.p_vaddr + (curr_off - start_off)
+        new_phdr.p_paddr = phdr.p_paddr + (curr_off - start_off)
+        new_phdr.p_filesz = size
+        new_phdr.p_memsz = size
+
+        if curr_off == start_off
+          phdr.assign(new_phdr)
+        else
+          new_phdrs << new_phdr
+        end
+
+        curr_off += size
+      end
+
+      new_phdrs
+    end
+
+    def sections_at_aligned_offset(offset)
+      @sections.each do |sec|
+        shdr = sec.header
+
+        aligned_offset = Helper.alignup(offset, shdr.sh_addralign)
+        next if shdr.sh_offset != aligned_offset
+
+        yield sec
+      end
+    end
+
+    def shift_sections(shift, start_offset)
+      ehdr.e_shoff += shift if ehdr.e_shoff >= start_offset
 
       @sections.each_with_index do |sec, i|
         next if i.zero? # dont touch NULL section
 
         shdr = sec.header
+        next if shdr.sh_offset < start_offset
+
         shdr.sh_offset += shift
       end
+    end
 
-      @segments.each do |seg|
+    def shift_segment_offset(phdr, shift)
+      phdr.p_offset += shift
+      phdr.p_align = page_size if phdr.p_align != 0 && (phdr.p_vaddr - phdr.p_offset) % phdr.p_align != 0
+    end
+
+    def shift_segment_virtual_address(phdr, shift)
+      phdr.p_paddr -= shift if phdr.p_paddr > shift
+      phdr.p_vaddr -= shift if phdr.p_vaddr > shift
+    end
+
+    def shift_segments(shift, start_offset)
+      split_index = -1
+      split_shift = 0
+
+      @segments.each_with_index do |seg, idx|
         phdr = seg.header
-        phdr.p_offset += shift
-        phdr.p_align = page_size if phdr.p_align != 0 && (phdr.p_vaddr - phdr.p_offset) % phdr.p_align != 0
+        p_start = phdr.p_offset
+
+        if p_start <= start_offset && p_start + phdr.p_filesz > start_offset &&
+           phdr.p_type == ELFTools::Constants::PT_LOAD
+          raise PatchError, "split_index(#{split_index}) != -1" if split_index != -1
+
+          split_index = idx
+          split_shift = start_offset - p_start
+
+          phdr.p_offset = start_offset
+          phdr.p_memsz -= split_shift
+          phdr.p_filesz -= split_shift
+          phdr.p_paddr += split_shift
+          phdr.p_vaddr += split_shift
+
+          p_start = start_offset
+        end
+
+        if p_start >= start_offset
```

Comment:
Shouldn't this be '>' ?

Reply:
start_offset is the start of the first section we want to shift, so this should be correct.

In the case above where I do `p_start = start_offset`,  what we're doing is first shortening that segment to now start at the start_offset, and then shifting the whole thing by the given amount.

The gap created is long enough for the new segment created in `shift_file`.

Reply:
I see, thanks!